src/chain/chain.rs

```diff
@@ -996,29 +996,44 @@ impl Chain {
         let is_connected = |target_name: &XorName| connected_peers.contains(&target_name);
 
         let candidates = |target_name: &XorName| {
-            let mut filtered_sections =
+            let filtered_sections =
                 self.closest_sections(target_name)
                     .into_iter()
-                    .map(|(_, members)| {
+                    .map(|(prefix, members)| {
                         (
+                            prefix,
                             members.len(),
                             members.into_iter().filter(is_connected).collect::<Vec<_>>(),
                         )
                     });
-            let best_section = filtered_sections
-                .find(|(len, connected)| connected.len() * 3 >= *len)
-                .ok_or(Error::CannotRoute);
-            best_section.map(|(len, connected)| {
-                (
-                    len,
-                    connected
-                        .into_iter()
-                        .sorted_by(|lhs, rhs| target_name.cmp_distance(lhs, rhs)),
-                )
-            })
+
+            let mut dg_size = 0;
+            let mut node_to_send = Vec::new();
+            for (idx, (prefix, len, connected)) in filtered_sections.enumerate() {
+                node_to_send.extend(connected.into_iter());
+                dg_size = delivery_group_size(len);
+
+                if &prefix == self.our_prefix() {
+                    // Send to all connected nodes so the connected node can forward the message
+                    node_to_send.retain(|&x| x != *self.our_id().name());
+                    dg_size = node_to_send.len();
+                    break;
+                }
+                if idx == 0 && node_to_send.len() >= dg_size {
+                    // can deliver to enough of the closest section
+                    break;
+                }
+            }
+            node_to_send.sort_by(|lhs, rhs| target_name.cmp_distance(lhs, rhs));
+
+            if dg_size > 0 && dg_size <= node_to_send.len() {
```

Comment:
in case of idx==0, we break at `dg_size >= node_to_send.len()`
however here we are looking for `dg_size <= node_to_send.len()`,
doesn't it mean only the case of `dg_size == node_to_send.len()` will pass ?

Reply:
```
dg_size = 3
node_to_send.len() = 8
idx = 0

if idx == 0 && node_to_send.len() >= dg_size   --> This is true, so break
if dg_size > 0 && dg_size <= node_to_send.len()  --> this is true so return Ok.
```

Reply:
ah, ok, the position of `dg_size` is flipped
maybe make it appear at the same front position?